### PR TITLE
Client: make fetcher more typesafe

### PR DESCRIPTION
**Pull request by jochem-brouwer**
This PR is very WIP. It currently makes the Fetcher a bit more typesafe, I am running into some type problems now with the HeaderFetcher and will need to do a bit more research what the types are intended to be under the hood and possibly have to introduce a third type on Fetcher.ts.

All `any` is now removed in Fetcher/BlockFetcher, need to fix HeaderFetcher, this has some nasty types which don't work with the type system I setup.

### CODE REVIEW

File: packages/client/lib/net/protocol/ethprotocol.ts [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { BN, bufferToInt } from 'ethereumjs-util'
 import { Block, BlockHeader, BlockHeaderBuffer } from '@ethereumjs/block'
 import { Chain } from './../../blockchain'
 import { Message, Protocol, ProtocolOptions } from './protocol'
+import { BlockBodyBuffer } from '../../../../block/dist'
```
Comment (ryanio) [inline packages/client/lib/net/protocol/ethprotocol.ts:5]:
```suggestion
import { BlockBodyBuffer } from '@ethereumjs/block'
```

Comment (jochem-brouwer) [reply packages/client/lib/net/protocol/ethprotocol.ts:5]: Ah, Visual Studio did this for me, nice catch 👍

File: packages/client/lib/sync/fetcher/headerfetcher.ts [hunk 2]
```diff
@@ -28,13 +33,17 @@ export class HeaderFetcher extends BlockFetcher {
...
       // we reached our request limit. try with a different peer.
-      return false
+      return undefined
+    }
+    const response = await (peer!.les as LesProtocolMethods).getBlockHeaders({ block: task.first, max: task.count })
+    if (response) {
+      return response.headers
     }
-    return (peer.les as LesProtocolMethods).getBlockHeaders({ block: task.first, max: task.count })
+    return undefined
```
Comment (ryanio) [inline packages/client/lib/sync/fetcher/headerfetcher.ts:46]: I think here you can just do:

```return response?.headers```

or if `response` not an optional value:

```return response.headers ?? undefined```

File: packages/client/lib/sync/fetcher/headerfetcher.ts [hunk 3]
```diff
@@ -28,13 +33,17 @@ export class HeaderFetcher extends BlockFetcher {
...
    * @param job
    */
-  async request(job: any) {
+  async request(job: Job<JobTask, BlockHeaderResult>) {
     const { task, peer } = job
-    if (this.flow.maxRequestCount(peer, 'GetBlockHeaders') < this.maxPerRequest) {
+    if (this.flow.maxRequestCount(peer!, 'GetBlockHeaders') < this.maxPerRequest) {
       // we reached our request limit. try with a different peer.
-      return false
+      return undefined
```
Comment (ryanio) [inline packages/client/lib/sync/fetcher/headerfetcher.ts:40]: is `return` and `return undefined` the same?

File: packages/client/lib/types.ts [hunk 4]
```diff
@@ -10,3 +12,36 @@ export interface Bootnode {
   port: number
 }
 export type BootnodeLike = string | string[] | Bootnode | Bootnode[]
+
+// QHeap types. @types/qheap does not exist, so we create a custom interface here.
```
Comment (ryanio) [inline packages/client/lib/types.ts:16]: nice. another option here is to make a `types/qheap.d.ts` file and then include it in the top of the file you import qheap with `/// <reference path="../types/qheap.d.ts" />`

Comment (ryanio) [reply packages/client/lib/types.ts:16]: in some cases i find it also needs to fed into the tsconfig [`typeRoots`](https://www.typescriptlang.org/tsconfig#typeRoots) if done this way ^

Comment (jochem-brouwer) [reply packages/client/lib/types.ts:14]: Thanks, I did not know that this was a feature. I will keep it in `types.ts` for now, though.

File: packages/client/lib/sync/fetcher/headerfetcher.ts (hunk 3 above)
Comment (jochem-brouwer) [reply packages/client/lib/sync/fetcher/headerfetcher.ts]: Yep, fixed.

File: packages/client/lib/sync/fetcher/headerfetcher.ts (hunk 2 above)
Comment (jochem-brouwer) [reply packages/client/lib/sync/fetcher/headerfetcher.ts]: Don't know why I wrote this since I can just return the result 😅 . Fixed.

**Comment by holgerd77** [comment]
Hi @jochem-brouwer, merry Christmas 🎄 😄 , without having had a closer look at the structure yet a first question: what is the reason for this separation of `BlockFetcherBase` and `Fetcher` couldn't this be included in the `Fetcher` abstract class directly without introducing an extra intermediary layer?

**Comment by holgerd77** [comment]
(that said, I might not have that much time for reviews during the next days...)

**Comment by jochem-brouwer** [comment]
Hi @holgerd77, Merry Christmas to you too 😄 🎄 . 

The motivation to create `BlockFetcherBase`, is because we make `Fetcher` abstract - `Fetcher` does not know what types it is going to handle. `BlockFetcherBase` includes the methods which are shared by both `BlockFetcher` and `HeaderFetcher` (they share some methods). By making `Fetcher` type-agnostic, this leaves us a lot of freedom to introduce new `Fetcher` classes, where we don't have to care about the "fetch" logic (which is in `Fetcher`), we only have to implement te abstract method. An example could be a `NodeFetcher`, to fetch nodes of the trie, which would definitely have different types than a `BlockFetcher`.

**Comment by holgerd77** [comment]
Rebased this.

**Review by holgerd77 — APPROVED** [review]
Spend the whole morning with rebasing #1017 on top of this PR. I've got it now and will merge here without further review, I am otherwise too afraid that we won't get this together anymore. 😜

I will directly afterwards push a new branch and open a PR with a rebased #1017, let's please do final fixes there (as well as eventual future rebase comments from this PR) and then do a relatively quick merge, this is otherwise getting to extensive.

**Review by holgerd77 — COMMENTED** [review]
Ok, had a post-merge review here, this looks good and I think merge was already generally justified, great job Jochem! 😄  👍  I still somewhat have my problems to read the new types here - I am still not so familiar with this object-centered angle brackets notation - but I think I will get into it when further working with the Fetcher classes.

**Comment by jochem-brouwer** [comment]
Hi @holgerd77, sorry for the late reply, these angle brackets have to do with TypeScript and do not exist in JavaScript. They have to do with passing on certain types to your object.

A great way to think about it would be to think of it as how you create Array types in TypeScript: you'd use `number[]` for instance to create a number-typed array. This is essentially a special case of angle brackets: there exists a type `Array<t>` where `number[]` is essentially the same as `Array<number>`.

**Pull request merged**